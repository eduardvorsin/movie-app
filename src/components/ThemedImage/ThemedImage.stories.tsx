import type { Meta, StoryObj } from '@storybook/react';
import ThemedImage from './ThemedImage';

const meta: Meta<typeof ThemedImage> = {
	title: 'UI/ThemedImage',
	tags: ['autodocs'],
	argTypes: {
		className: {
			description: 'Additional classes for the component'
		},
		testId: {
			description: 'Id for testing the component',
		},
		src: {
			description: 'The path for the image',
		},
		darkSrc: {
			description: 'The path for the image when the dark theme of the site is enabled',
		},
		fallback: {
			description: 'The path to the backup image if the main image fails to load',
		},
		alt: {
			description: 'Alternative text for the image',
		},
		placeholder: {
			control: false,
			description: 'The placeholder that will be shown when the image is uploaded',
		},
		fill: {
			description: 'Makes the image absolute positioning and filling the entire width and height',
		},
		width: {
			description: 'Image Width',
		},
		height: {
			description: 'Image Height',
		},
		showSkeleton: {
			description: 'Allows you to determine whether the skeleton will be displayed when the image is loaded or not',
		},
	},
	component: ThemedImage,
	parameters: {
		docs: {
			description: {
				component: 'ThemedImage is an image component with the ability to customize the theme, providing flexibility in integrating images into various design schemes of web applications.',
			},
		},
	},
}

export default meta;
type Story = StoryObj<typeof ThemedImage>;

export const Default: Story = {
	args: {
		width: 250,
		height: 250,
		src: '/assets/images/collection-marvel.webp'
	},
	parameters: {
		docs: {
			description: {
				story: 'The default state of the component',
			},
		},
	},
};

export const DarkSrc: Story = {
	args: {
		width: 250,
		height: 250,
		src: '/assets/images/collection-marvel.webp',
		darkSrc: '/assets/images-collection-dc.webp',
	},
	parameters: {
		docs: {
			description: {
				story: 'Shows a different image if the theme is dark',
			},
		},
	},
};

export const Fallback: Story = {
	args: {
		src: '',
		width: 250,
		height: 250,
		fallback: {
			light: '/assets/images/movie-card-placeholder-l-h.svg',
			dark: '/assets/images/movie-card-placeholder-d-h.svg'
		},
	},
	parameters: {
		docs: {
			description: {
				story: 'Shows a backup image when the main image is not loading',
			},
		},
	},
};