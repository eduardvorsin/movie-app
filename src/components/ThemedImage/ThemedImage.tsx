'use client';
import { Theme, useTheme } from '@/context/ThemeProvider/ThemeProvider';
import Image, { ImageProps } from 'next/image';

type Props = {
	src: Record<Theme, string>;
	fallback?: Record<Theme, string>;
} & Omit<ImageProps, 'src'>;
export default function ThemedImage({
	src,
	fallback,
	...props
}: Props) {
	const theme = useTheme();

	let currentSrc: string;
	if (src[theme].length) {
		currentSrc = src[theme];
	} else {
		const inverseTheme = theme === 'light' ? 'dark' : 'light';
		console.log('inverseTheme', inverseTheme);
		currentSrc = fallback?.[theme] ?? src[inverseTheme];
	}

	return (
		// eslint-disable-next-line jsx-a11y/alt-text
		<Image
			src={currentSrc}
			{...props}
		/>
	);
};