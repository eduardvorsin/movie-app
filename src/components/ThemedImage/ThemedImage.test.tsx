import { render, screen } from "@testing-library/react";
import ThemedImage from "./ThemedImage";

describe('ThemedImage tests', () => {
	it('is rendered correctly', () => {
		render(
			<ThemedImage
				width={100}
				height={100}
				alt='test-themed-image'
				src='/assets/images/movie-card-placeholder-l-v.svg'
				testId='test-themed-image'
			/>
		);

		expect(screen.getByAltText<HTMLImageElement>('test-themed-image')).toBeInTheDocument();
	});

	it('if src.length is 0, then the image from the fallback property is rendered', () => {
		render(
			<ThemedImage
				width={100}
				height={100}
				alt='test-themed-image'
				src=''
				fallback={{
					light: '/assets/images/movie-card-placeholder-l-v.svg',
					dark: '/assets/images/movie-card-placeholder-d-v.svg',
				}}
				testId='test-themed-image'
			/>
		);

		expect(screen.getByAltText<HTMLImageElement>('test-themed-image').src).toMatch(/movie-card-placeholder/i);
	});

	it('if the dark theme is enabled, the image from the darkSrc property is rendered', () => {
		render(
			<ThemedImage
				width={100}
				height={100}
				alt='test-themed-image'
				src=''
				darkSrc='/assets/images/movie-card-placeholder-d-v.svg'
				testId='test-themed-image'
			/>
		);

		expect(screen.getByAltText<HTMLImageElement>('test-themed-image').src).toMatch(/movie-card-placeholder/i);
	})

	it('is a basic snapshot', () => {
		render(
			<ThemedImage
				width={100}
				height={100}
				alt='test-themed-image'
				src='/assets/images/movie-card-placeholder-l-v.svg'
				testId='test-themed-image'
			/>
		);

		expect(screen.getByAltText<HTMLImageElement>('test-themed-image')).toMatchSnapshot();
	});

	it('is a snapshot with a fallback', () => {
		render(
			<ThemedImage
				width={100}
				height={100}
				alt='test-themed-image'
				src=''
				fallback={{
					light: '/assets/images/movie-card-placeholder-l-v.svg',
					dark: '/assets/images/movie-card-placeholder-d-v.svg',
				}}
			/>
		);

		expect(screen.getByAltText<HTMLImageElement>('test-themed-image')).toMatchSnapshot();
	});

	it('is a snapshot with showSkeleton set to true', () => {
		render(
			<ThemedImage
				width={100}
				height={100}
				alt='test-themed-image'
				src='/'
				showSkeleton
				fallback={{
					light: '/assets/images/movie-card-placeholder-l-v.svg',
					dark: '/assets/images/movie-card-placeholder-d-v.svg',
				}}
			/>
		);

		expect(screen.getByAltText<HTMLImageElement>('test-themed-image')).toMatchSnapshot();
	});

	it('is a snapshot with darkSrc', () => {
		render(
			<ThemedImage
				width={100}
				height={100}
				alt='test-themed-image'
				src='/assets/images/movie-card-placeholder-d-v.svg'
				darkSrc='/assets/images/movie-card-placeholder-d-v.svg'
				testId='test-themed-image'
			/>
		);

		expect(screen.getByAltText<HTMLImageElement>('test-themed-image')).toMatchSnapshot();
	});
});