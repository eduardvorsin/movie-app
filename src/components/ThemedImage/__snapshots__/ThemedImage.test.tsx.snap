// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemedImage tests is a basic snapshot 1`] = `
<img
  alt="test-themed-image"
  data-nimg="1"
  data-testid="test-themed-image"
  decoding="async"
  height="100"
  loading="lazy"
  src="http://localhost/assets/images/movie-card-placeholder-l-v.svg"
  style="color: transparent;"
  width="100"
/>
`;

exports[`ThemedImage tests is a snapshot with a fallback 1`] = `
<img
  alt="test-themed-image"
  data-nimg="1"
  decoding="async"
  height="100"
  loading="lazy"
  src="http://localhost/assets/images/movie-card-placeholder-d-v.svg"
  style="color: transparent;"
  width="100"
/>
`;

exports[`ThemedImage tests is a snapshot with darkSrc 1`] = `
<img
  alt="test-themed-image"
  data-nimg="1"
  data-testid="test-themed-image"
  decoding="async"
  height="100"
  loading="lazy"
  src="http://localhost/assets/images/movie-card-placeholder-d-v.svg"
  style="color: transparent;"
  width="100"
/>
`;

exports[`ThemedImage tests is a snapshot with showSkeleton set to true 1`] = `
<img
  alt="test-themed-image"
  data-nimg="fill"
  decoding="async"
  loading="lazy"
  sizes="100vw"
  src="http://localhost/_next/image?url=%2F&w=3840&q=75"
  srcset="/_next/image?url=%2F&w=640&q=75 640w, /_next/image?url=%2F&w=750&q=75 750w, /_next/image?url=%2F&w=828&q=75 828w, /_next/image?url=%2F&w=1080&q=75 1080w, /_next/image?url=%2F&w=1200&q=75 1200w, /_next/image?url=%2F&w=1920&q=75 1920w, /_next/image?url=%2F&w=2048&q=75 2048w, /_next/image?url=%2F&w=3840&q=75 3840w"
  style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
/>
`;
